parser/parser_test.go

```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	expectedType := expected.item.Type()
+	if actualType == expectedType {
+		return true, ""
+	}
+	return false, actual.name + " is of type " + actualType.String() + " whereas " + expected.name + " is of type " + expectedType.String()
+}
+
+func deepCompareObjects(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	depth++
+	if depth > 50 {
+		panic(fmt.Sprintf("Detected probable infinite loop comparing %s (%s) and %s (%s). This diff checker doesn't yet support data structures having cycles.",
+			actual.name,
+			actual.item,
+			expected.name,
+			expected.item))
+	}
+
+	equal, diff = compareObjectsClasses(actual, expected, depth)
+	if !equal {
+		return equal, diff
+	}
+
+	switch actual.item.Kind() {
+	case reflect.Array, reflect.Slice:
+		return deepCompareArrays(actual, expected, depth)
+	case reflect.Map:
+		return deepCompareMaps(actual, expected, depth)
+	case reflect.Struct:
+		return deepCompareStructs(actual, expected, depth)
+	case reflect.Ptr, reflect.Interface:
+		return deepComparePointers(actual, expected, depth)
+	default:
+		return comparePrimitives(actual, expected, depth)
+	}
+
+}
+
+func deepComparePointers(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	if actual.item.IsNil() || expected.item.IsNil() {
+		if actual.item.IsNil() && expected.item.IsNil() {
+			return true, ""
+		} else {
+
+			if actual.item.IsNil() {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", actual.name, expected.name, expected.item)
+			} else {
+				return false, fmt.Sprintf("%s is nil whereas %s is %#v", expected.name, actual.name, actual.item)
 			}
+		}
+
+	}
+	return deepCompareObjects(namedItem{actual.name, actual.item.Elem()},
+		namedItem{expected.name, expected.item.Elem()},
+		depth)
+}
+
+func comparePrimitives(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	// TODO make this less hacky
+	var a, b interface{}
+	if actual.item.Kind() != expected.item.Kind() {
+		return false, fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind())
+	}
+	switch actual.item.Kind() {
+	case reflect.String:
+		a = actual.item.String()
+		b = expected.item.String()
+	case reflect.Bool:
+		a = actual.item.Bool()
+		b = expected.item.Bool()
+	case reflect.Int:
+		a = actual.item.Int()
+		b = expected.item.Int()
+	default:
+		panic(fmt.Sprintf("unrecognized object types, %s (%#v) and %s (%#v)", actual.name, actual.item, expected.name, expected.item))
+	}
+	if reflect.DeepEqual(a, b) {
+		return true, ""
+	} else {
+		return false, fmt.Sprintf("%s = %#v whereas %s = %#v", actual.name, actual.item, expected.name, expected.item)
+	}
+
+}
+
+func deepCompareMaps(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aKeys := a.MapKeys()
+	bKeys := b.MapKeys()
+	for _, aKey := range aKeys {
+		if !(isIn(aKey, bKeys)) {
```

Comment:
Yeah exactly; the test doesn't confirm that the pointers themselves are correct, only that they have a corresponding value in the other map.

Do you mean that we want to fix deepCompare for the following case?

////
var a and b, deeply equal to each other
var c and d, deeply equal to each other

map1{a:true, b:true, c:true}
map2{a:true, c:true, d:true}

?

for which deepCompare would return true when really we want false